File: src/v2/config/resolver.ts
/**
 * Environment Variable Resolver
 *
 * Resolves ${VAR_NAME} patterns in config values to actual environment variable values.
 */

// =============================================================================
// TYPES
// =============================================================================

export interface ResolverOptions {
  throwOnMissing?: boolean;
  prefix?: string;
}

export interface ResolverResult {
  value: string;
  resolved: boolean;
  missing?: string[];
}

// =============================================================================
// RESOLVER
// =============================================================================

const ENV_VAR_PATTERN = /\$\{([^}]+)\}/g;

/**
 * Resolve environment variables in a string
 * Supports ${VAR_NAME} syntax
 *
 * @example
 * resolveEnvVars('https://${HOST}:${PORT}')
 * // Returns 'https://localhost:8080' if HOST=localhost, PORT=8080
 */
export function resolveEnvVars(
  value: string,
  options: ResolverOptions = {}
): ResolverResult {
  const { throwOnMissing = false, prefix = '' } = options;
  const missing: string[] = [];
  let resolved = false;

  const result = value.replace(ENV_VAR_PATTERN, (match, varName) => {
    const fullVarName = prefix ? `${prefix}${varName}` : varName;
    const envValue = process.env[fullVarName];

    if (envValue === undefined) {
      missing.push(fullVarName);
      if (throwOnMissing) {
        throw new Error(`Missing environment variable: ${fullVarName}`);
      }
      return match; // Keep original if not found
    }

    resolved = true;
    return envValue;
  });

  return { value: result, resolved, missing: missing.length > 0 ? missing : undefined };
}

/**
 * Check if a string contains environment variable references
 */
export function hasEnvVars(value: string): boolean {
  return ENV_VAR_PATTERN.test(value);
}

/**
 * Extract environment variable names from a string
 */
export function extractEnvVarNames(value: string): string[] {
  const names: string[] = [];
  let match;

  // Reset lastIndex for global regex
  ENV_VAR_PATTERN.lastIndex = 0;

  while ((match = ENV_VAR_PATTERN.exec(value)) !== null) {
    names.push(match[1]);
  }

  return names;
}

/**
 * Resolve all environment variables in an object recursively
 */
export function resolveEnvVarsInObject<T extends object>(
  obj: T,
  options: ResolverOptions = {}
): { value: T; missing: string[] } {
  const allMissing: string[] = [];

  function resolveValue(val: unknown): unknown {
    if (typeof val === 'string') {
      const result = resolveEnvVars(val, { ...options, throwOnMissing: false });
      if (result.missing) {
        allMissing.push(...result.missing);
      }
      return result.value;
    }

    if (Array.isArray(val)) {
      return val.map(resolveValue);
    }

    if (val && typeof val === 'object') {
      const resolved: Record<string, unknown> = {};
      for (const [key, value] of Object.entries(val)) {
        resolved[key] = resolveValue(value);
      }
      return resolved;
    }

    return val;
  }

  const resolved = resolveValue(obj) as T;

  if (options.throwOnMissing && allMissing.length > 0) {
    throw new Error(`Missing environment variables: ${allMissing.join(', ')}`);
  }

  return { value: resolved, missing: allMissing };
}

/**
 * Validate that all required environment variables are set
 */
export function validateEnvVars(
  requiredVars: string[]
): { valid: boolean; missing: string[] } {
  const missing = requiredVars.filter(name => !process.env[name]);
  return { valid: missing.length === 0, missing };
}

File: src/foundry_rules/config/resolver.py
"""
Environment variable resolver.

Handles ${VAR_NAME} pattern resolution in config values.
"""

import os
import re
from typing import Optional


# Pattern to match ${VAR_NAME}
ENV_VAR_PATTERN = re.compile(r"\$\{([^}]+)\}")


def has_env_vars(value: str) -> bool:
    """Check if a string contains environment variable references."""
    return bool(ENV_VAR_PATTERN.search(value))


def extract_env_var_names(value: str) -> list[str]:
    """Extract all environment variable names from a string."""
    return ENV_VAR_PATTERN.findall(value)


def resolve_env_vars(value: str) -> tuple[str, list[str]]:
    """
    Resolve environment variables in a string.

    Args:
        value: String potentially containing ${VAR_NAME} patterns

    Returns:
        Tuple of (resolved_value, list_of_missing_vars)
    """
    missing: list[str] = []

    def replacer(match: re.Match) -> str:
        var_name = match.group(1)
        env_value = os.environ.get(var_name)
        if env_value is None:
            missing.append(var_name)
            return match.group(0)  # Keep original if not found
        return env_value

    resolved = ENV_VAR_PATTERN.sub(replacer, value)
    return resolved, missing


def validate_env_vars(var_names: list[str]) -> tuple[bool, list[str]]:
    """
    Check if environment variables are set.

    Args:
        var_names: List of environment variable names to check

    Returns:
        Tuple of (all_valid, list_of_missing_vars)
    """
    missing = [name for name in var_names if os.environ.get(name) is None]
    return len(missing) == 0, missing


def get_env_var(name: str, default: Optional[str] = None) -> Optional[str]:
    """Get an environment variable value."""
    return os.environ.get(name, default)
